File: 02 백준/01 자료구조/Python/Prob19583.py

"""
title  : 19583. 싸이버개강총회 (Silver 2)
time   : 100ms     196ms
memory : 43548KB   123820KB
"""

import sys

input = sys.stdin.readline

def solution():
    S, E, Q = input().split()       # 개강총회를 시작한 시간 S, 끝낸 시간 E, 스트리밍을 끝낸 시간 Q

    people = set()
    answer = 0

    while True:
        try:
            time, name = input().split()

            if time <= S:
                people.add(name)
            elif E <= time <= Q:
                if name in people:
                    people.remove(name)
                    answer += 1
        except:
            break

    return answer

print(solution())

File: 02 백준/01 자료구조/Java/src/Prob19583.java

/**
* title  : 19583. 싸이버개강총회 (Silver 2)
* time   : 500ms
* memory : 69332KB
*/

import java.io.*;
import java.util.*;

public class Prob19583 {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int S = timeToInt(st.nextToken());
        int E = timeToInt(st.nextToken());
        int Q = timeToInt(st.nextToken());

        HashMap<String, Integer> streaming = new HashMap<>();
        int answer = 0;

        String student;
        while ((student = br.readLine()) != null) {
            // if (student.length() < 3) break;

            st = new StringTokenizer(student);
            int now = timeToInt(st.nextToken());
            String name = st.nextToken();

            if (now <= S) {
                streaming.put(name, 1);
            } else if (E <= now && now <= Q) {
                if (streaming.getOrDefault(name, -1) == 1) {
                    streaming.replace(name, 0);
                    answer++;
                }
            }
        }

        System.out.println(answer);
    }

    public static int timeToInt(String time) {
        String[] temp = time.split(":");
        return Integer.parseInt(temp[0]) * 60 + Integer.parseInt(temp[1]);
    }
}
